src/main/webapp/go.js

```diff
@@ -0,0 +1,143 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const GO_URL = '/go-data';
+
+const INVISIBLE_CLASS = 'slide-invisible';
+const NAME_URL = '/name-data';
+const START_ID = 'start-button';
+const PROCEED_ID = 'proceed-button';
+const RIDDLE_ID = 'riddle-area';
+const INDEX_PARAM = 'new-index';
+const FINAL_MSSG = 'Congrats, you\'ve finished the hunt!';
+
+const riddleArr = [];
+let destIndex;
+const destArr = [];
+
+window.onload = getHunt();
+
+/**
+ * Retrieves scavenger hunt data, and updates to the current destination
+ * to reflect the current state of the hunt.
+ */
+function getHunt() {
+  fetch(GO_URL).then((response) => response.json()).then((mssg) => {
+    destIndex = mssg.index;
+    for (let i = 0; i < mssg.items.length; i++) {
+      riddleArr.push(mssg.items[i].riddle.puzzle);
+      destArr.push(mssg.items[i].name);
+    }
+    updateToCurrentState(destIndex);
+  });
+}
+
+/**
+ * Updates the hunt to the current destination that the user is on.
+ *
+ * @param {int} index The index of the destination that the user needs
+ * to find.
+ */
+function updateToCurrentState(index) {
+  if (index >= 0) {
+    hideStartButton();
+    changeRiddleMessage(riddleArr[index]);
+    getDestName();
+  }
+  sendIndexToServlet(index);
+  hideProceedButton();
+}
+
+/**
+ * Retrieves the string the user submitted as the destination name.
+ */
+function getDestName() {
+  fetch(NAME_URL).then((response) => response.json()).then((mssg) => {
```

Comment:
It looks like in this design, the user submits a guess which is then stored on the server.  When this method is called in updateToCurrentState, it retrieves the stored guess to check it.

I am wondering if this needs to be stored on the server at all? Will it need to be stored on the server in the future?  Instead, you could just check that the value of the input field matches destArr[destIndex], which could all be done client-side.

Reply:
I think it's fine to send to the server if it's going to be used for the MVP.  It would be helpful to comment that that's the intention so that it's clear why.

Reply:
I think this function might also be renamed checkCorrectDestination or something since that better describes what is happening.  And now that I think about it, this method has a side-effect:  if the name matches, it shows the proceed button.  It not, it does nothing.  But to the caller of this method, updateToCurrentState, it doesn't appear that way.

It would be clearer for checkCorrectDestination to simply return a boolean (the method would need to be async in this case, since it's doing a fetch).  Then in updateToCurrentState, you would know whether to show the proceed button or not.  I think this will make updateToCurrentState easier to understand since the logic flow is outlined better.